src/core/lib/channel/connected_channel.cc

```diff
@@ -745,95 +563,116 @@ ArenaPromise<ServerMetadataHandle> MakeServerCallPromise(
+      TrySeq(GetContext<BatchBuilder>()->ReceiveClientInitialMetadata(
+                 stream->batch_target()),
+             [next_promise_factory = std::move(next_promise_factory),
+              server_to_client_empty = std::move(server_to_client_empty),
+              call_data](ClientMetadataHandle client_initial_metadata) {
+               auto call_promise = next_promise_factory(CallArgs{
+                   std::move(client_initial_metadata),
+                   ClientInitialMetadataOutstandingToken::Empty(),
+                   &call_data->server_initial_metadata.sender,
+                   &call_data->client_to_server.receiver,
+                   &call_data->server_to_client.sender,
+               });
+               return Race(call_data->failure_latch.Wait(),
+                           [call_promise = std::move(call_promise),
+                            server_to_client_empty =
+                                std::move(server_to_client_empty)]() mutable
+                           -> Poll<ServerMetadataHandle> {
+                             // TODO(ctiller): this is deeply weird and we need
+                             // to clean this up.
+                             //
+                             // The following few lines check to ensure that
+                             // there's no message currently pending in the
+                             // outgoing message queue, and if (and only if)
+                             // that's true decides to poll the main promise to
+                             // see if there's a result.
+                             //
+                             // This essentially introduces a polling priority
+                             // scheme that makes the current promise structure
+                             // work out the way we want when talking to
+                             // transports.
+                             //
+                             // The problem is that transports are going to need
+                             // to replicate this structure when they convert to
+                             // promises, and that becomes troubling as we'll be
+                             // replicating weird throughout the stack.
+                             //
+                             // Instead we likely need to change the way we're
+                             // composing promises through the stack.
+                             //
+                             // Proposed is to change filters from a promise
+                             // that takes ClientInitialMetadata and returns
+                             // ServerTrailingMetadata with three pipes for
+                             // ServerInitialMetadata and
+                             // ClientToServerMessages, ServerToClientMessages.
+                             // Instead we'll have five pipes, moving
+                             // ClientInitialMetadata and ServerTrailingMetadata
```

Comment:
If server trailing metadata is a pipe, doesn't that eliminate one of the big advantages of the promise stack, which is that the promise always resolves to the call's status?  And doesn't it reintroduce the possibility that client initial metadata is not immediately available when the call starts?

Reply:
we'll keep those ordering constraints in the leaf filters (call.cc, and the terminators), but move the ordering detail out of the filters: they can rely on it, but don't need to implement it -- in that way we can't have problems if one filter implements the ordering slightly wrong.